File: src/main/java/ro/sopa/statistifier/api/client/LastFMClient.java
package ro.sopa.statistifier.api.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;
import ro.sopa.statistifier.api.model.ListeningHistory;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

@Component
public class LastFMClient {

    private ObjectMapper objectMapper = new ObjectMapper();

    private static final int PAGE_SIZE = 200;

    private static final String URL = "https://ws.audioscrobbler.com/2.0/?method=user.getrecenttracks&user=%s&api_key=%s&format=json&page=%s&limit=%s";

    public ListeningHistory callRecentTracks(String apiKey, String username, int page) {

        HttpClient client = HttpClient.newHttpClient();

        String url = String.format(URL, username, apiKey, page, PAGE_SIZE);

        HttpRequest request = null;
        try {
            request = HttpRequest.newBuilder()
                    .uri(new URI(url))
                    .POST(HttpRequest.BodyPublishers.noBody())
                    .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() == 200) {
                String responseBody = response.body().toString();
                try {
                    return objectMapper.readValue(responseBody, ListeningHistory.class);
                } catch (JsonProcessingException e) {
                    throw new RuntimeException("Can't deserialise response!");
                }

            } else {
                throw new RuntimeException("Can't call endpoint, received " + response.statusCode() + ". Mesage is " + response.body().toString());
            }

        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }

    }
}

File: src/main/kotlin/ro/sopa/lastfm/api/client/LastFMClient.kt
package ro.sopa.lastfm.api.client

import com.fasterxml.jackson.core.JsonProcessingException
import com.google.gson.Gson
import org.springframework.stereotype.Component
import ro.sopa.lastfm.api.model.ListeningHistory
import ro.sopa.lastfm.api.model.correction.CorrectionResponse
import ro.sopa.lastfm.api.model.toptags.TopTagsResponse
import java.io.IOException
import java.net.URI
import java.net.URISyntaxException
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets

@Component
class LastFMClient {

    private val gson = Gson()

    private val pageSize: Int = 200

    private val recentTracksUrl = "https://ws.audioscrobbler.com/2.0/?method=user.getrecenttracks&user=%s&api_key=%s&format=json&page=%s&limit=%s"
    private val correctionsUrl = "http://ws.audioscrobbler.com/2.0/?method=artist.getcorrection&artist=%s&api_key=%s&format=json"
    private val topTagsUrl = "http://ws.audioscrobbler.com/2.0/?method=artist.gettoptags&artist=%s&api_key=%s&format=json"

    fun callRecentTracks(apiKey: String, username: String, page: Int): ListeningHistory {

        val client = HttpClient.newHttpClient()
        val url = String.format(recentTracksUrl, username, apiKey, page, pageSize)

        var request: HttpRequest?

        return try {
            request = HttpRequest.newBuilder()
                .uri(URI(url))
                .POST(HttpRequest.BodyPublishers.noBody())
                .build()
            val response = client.send(request, HttpResponse.BodyHandlers.ofString())
            if (response.statusCode() == 200) {
                val responseBody = response.body().toString()
                try {
                    gson.fromJson(responseBody, ListeningHistory::class.java)
                } catch (e: JsonProcessingException) {
                    throw RuntimeException("Can't deserialise response!")
                }
            } else {
                throw RuntimeException("Can't call endpoint, received " + response.statusCode() + ". Mesage is " + response.body().toString())
            }

        } catch (e: URISyntaxException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw RuntimeException(e)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }

    fun callCorrections(apiKey: String, artist: String): CorrectionResponse {

        val client = HttpClient.newHttpClient()
        val url = String.format(correctionsUrl, URLEncoder.encode(artist, StandardCharsets.UTF_8.toString()), apiKey)
        var request: HttpRequest?

        return try {
            request = HttpRequest.newBuilder()
                .uri(URI(url))
                .POST(HttpRequest.BodyPublishers.noBody())
                .build()
            val response = client.send(request, HttpResponse.BodyHandlers.ofString())
            if (response.statusCode() == 200) {
                val responseBody = response.body().toString()
                try {
                    return gson.fromJson(responseBody, CorrectionResponse::class.java)
                } catch (e: Exception) {
                    throw RuntimeException("Can't deserialise response!", e)
                }
            } else {
                throw RuntimeException("Can't call endpoint, received " + response.statusCode() + ". Mesage is " + response.body().toString())
            }

        } catch (e: URISyntaxException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw RuntimeException(e)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }

    fun callTopTags(apiKey: String, artist: String): TopTagsResponse {
        val client = HttpClient.newHttpClient()
        val url = String.format(topTagsUrl, URLEncoder.encode(artist, StandardCharsets.UTF_8.toString()), apiKey)
        var request: HttpRequest?

        return try {
            request = HttpRequest.newBuilder()
                .uri(URI(url))
                .POST(HttpRequest.BodyPublishers.noBody())
                .build()
            val response = client.send(request, HttpResponse.BodyHandlers.ofString())
            if (response.statusCode() == 200) {
                val responseBody = response.body().toString()
                try {
                    return gson.fromJson(responseBody, TopTagsResponse::class.java)
                } catch (e: Exception) {
                    throw RuntimeException("Can't deserialise response!", e)
                }
            } else {
                throw RuntimeException("Can't call endpoint, received " + response.statusCode() + ". Mesage is " + response.body().toString())
            }

        } catch (e: URISyntaxException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw RuntimeException(e)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
    }
}